src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationFilter.cs

```diff
@@ -0,0 +1,171 @@
+
+            _requirements = requirements;
+        }
+
+        /// <summary>
+        /// Called early in the filter pipeline to confirm request is authorized.
+        /// </summary>
+        /// <param name="context">The <see cref="T:Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext" />.</param>
+        public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
+        {
+            Guard.NotNull(context, nameof(context));
+            Guard.NotNull(context.HttpContext, nameof(context.HttpContext));
+            Guard.For<ArgumentException>(() => context.HttpContext.Connection is null, "Invalid action context given without any HTTP connection");
+            Guard.For<ArgumentException>(() => context.HttpContext.RequestServices is null, "Invalid action context given without any HTTP request services");
+
+            ILogger logger = 
+                context.HttpContext.RequestServices
+                       .GetService<ILoggerFactory>()
+                       ?.CreateLogger<CertificateAuthenticationFilter>() 
+                ?? (ILogger) NullLogger.Instance;
+
+            ISecretProvider userDefinedSecretProvider = 
+                context.HttpContext.RequestServices.GetService<ICachedSecretProvider>()
+                ?? context.HttpContext.RequestServices.GetService<ISecretProvider>();
+
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ICachedSecretProvider)} or {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            X509Certificate2 clientCertificate = context.HttpContext.Connection.ClientCertificate;
+            if (clientCertificate == null)
+            {
+                logger.LogWarning(
+                    "No client certificate was specified in the HTTP request while this authentication filter "
+                    + $"requires a certificate to validate on the {String.Join(", ", _requirements.Select(item => item.requirement))}");
+                
+                context.Result = new UnauthorizedResult();
+            }
+            else if (!await IsAllowedCertificate(clientCertificate, userDefinedSecretProvider, logger))
+            {
+                context.Result = new UnauthorizedResult();
+            }
+        }
+
+        private async Task<bool> IsAllowedCertificate(
+            X509Certificate2 clientCertificate, 
+            ISecretProvider provider,
+            ILogger logger)
+        {
+            var requirementValues = await Task.WhenAll(_requirements.Select(async item =>
+            {
+                string expected = await provider.Get(item.configurationKey);
```

Comment:
It would be better if all async methods are suffixed with `Async` IMO.  It makes it immediately obvious that the method is async and that you'd probably want to `await` it.  (But I believe this `ISecretProvider.Get` method comes from another arcus repo ?

Reply:
I think that too, but since all the Arcus async methods are not written this way; it would break consistency.

Reply:
I thought this was defined in the arcus coding guidlines which we defined back in the day but apparently not. https://codit.visualstudio.com/Arcus/_wiki/wikis/Arcus.wiki?wikiVersion=GBwikiMaster&pagePath=%2FGuidance%2FCoding%20Guidelines&pageId=10

I am on the same page on that perspective, however we need to see how our public methods are now and if we should align them.

Think a seperate issue for tracking this will be best option.

Reply:
Created https://github.com/arcus-azure/arcus/issues/28 for this